src/measure_theory/ae_eq_fun.lean

```diff
@@ -0,0 +1,329 @@
+          [add_comm_monoid γ] [semimodule K γ] [topological_semimodule K γ]
+
+protected def smul : K → (α →ₘ γ) → (α →ₘ γ) :=
+λ c f, comp (has_scalar.smul c) (measurable_smul measurable_id) f
+
+instance : has_scalar K (α →ₘ γ) := ⟨ae_eq_fun.smul⟩
+
+@[simp] lemma smul_mk (c : K) (f : α → γ) (hf) : c • (mk f hf) = mk (c • f) (measurable_smul hf) :=
+rfl
+
+variables [topological_add_monoid γ]
+
+instance : semimodule K (α →ₘ γ) :=
+{ one_smul  := by { rintros ⟨f, hf⟩, simp only [quot_mk_eq_mk, smul_mk, one_smul] },
+  mul_smul  :=
+    by { rintros x y ⟨f, hf⟩, simp only [quot_mk_eq_mk, smul_mk, mul_action.mul_smul x y f], refl },
+  smul_add  :=
+  begin
+    rintros x ⟨f, hf⟩ ⟨g, hg⟩, simp only [quot_mk_eq_mk, smul_mk, mk_add_mk],
+    congr, exact smul_add x f g
+  end,
+  smul_zero := by { intro x, simp only [zero_def, smul_mk], congr, exact smul_zero x },
+  add_smul  :=
+  begin
+    intros x y, rintro ⟨f, hf⟩, simp only [quot_mk_eq_mk, smul_mk, mk_add_mk], congr,
+    exact add_smul x y f
+  end,
+  zero_smul :=
+    by { rintro ⟨f, hf⟩, simp only [quot_mk_eq_mk, smul_mk, zero_def], congr, exact zero_smul K f }}
+
+end semimodule
+
+section vector_space
+
+variables {K : Type*} [discrete_field K] [topological_space K] [second_countable_topology K]
+variables {γ : Type*} [topological_space γ] [second_countable_topology γ] [add_comm_group γ]
+          [topological_add_group γ] [vector_space K γ] [topological_semimodule K γ]
+
+instance : vector_space K (α →ₘ γ) := { .. ae_eq_fun.semimodule }
+
+end vector_space
+
+open ennreal
+-- integral on ae_eq_fun
+def integral (f : α →ₘ nnreal) : ennreal :=
+quotient.lift_on f (λf, ∫⁻ a, f.1 a)
+  (assume ⟨f, hf⟩ ⟨g, hg⟩ eq, lintegral_congr_ae $ by simpa [coe_eq_coe])
+
+@[simp] lemma integral_mk (f : α → nnreal) (hf) : integral (mk f hf) = ∫⁻ a, f a :=
+rfl
+
+@[simp] lemma integral_zero : integral (0 : α →ₘ nnreal) = 0 :=
+lintegral_zero
+
+@[simp] lemma integral_eq_zero_iff (f : α →ₘ nnreal) : integral f = 0 ↔ f = 0 :=
+begin
+  rcases f with ⟨f, hf⟩,
+  refine iff.trans (lintegral_eq_zero_iff _) ⟨_, _⟩,
+  { exact measurable_coe.comp hf },
+  { assume h, simp only [coe_eq_zero] at h, exact quotient.sound h },
+  { assume h, simp only [coe_eq_zero], exact quotient.exact h }
+end
+
+lemma integral_add : ∀(f g : α →ₘ nnreal), integral (f + g) = integral f + integral g :=
+by rintros ⟨f⟩ ⟨g⟩; simp only [quot_mk_eq_mk, mk_add_mk, integral_mk, coe_add,
+   lintegral_add (measurable_coe.comp f.2) (measurable_coe.comp g.2)]
+
+lemma integral_le_integral {f g : α →ₘ nnreal} (h : f ≤ g) : integral f ≤ integral g :=
+begin
+  rcases f with ⟨f, hf⟩, rcases g with ⟨g, hg⟩,
+  simp only [quot_mk_eq_mk, integral_mk, mk_le_mk] at *,
+  refine lintegral_le_lintegral_ae _,
+  filter_upwards [h], simp
+end
+
+section metric
+
+variables {γ : Type*} [metric_space γ] [second_countable_topology γ]
+
+def comp_nndist (f g : α →ₘ γ) : α →ₘ nnreal := comp₂ nndist measurable_nndist' f g
```

Comment:
Here again, define instead `comp_edist` and define directly `edist` below avoiding `nnreal` altogether.